src/app.rs

```diff
@@ -336,9 +328,65 @@ impl eframe::App for MacroSolverApp {
             ui.add(StatsEdit::new(self.locale, &mut self.crafter_config));
         });
     }
+
+    fn save(&mut self, storage: &mut dyn eframe::Storage) {
+        eframe::set_value(storage, "LOCALE", &self.locale);
+        eframe::set_value(storage, "RECIPE_CONFIG", &self.recipe_config);
+        eframe::set_value(storage, "SELECTED_FOOD", &self.selected_food);
+        eframe::set_value(storage, "SELECTED_POTION", &self.selected_potion);
+        eframe::set_value(storage, "CRAFTER_CONFIG", &self.crafter_config);
+        eframe::set_value(storage, "SOLVER_CONFIG", &self.solver_config);
+        eframe::set_value(storage, "MACRO_VIEW_CONFIG", &self.macro_view_config);
+
+        eframe::set_value(storage, "CUSTOM_RECIPE", &self.custom_recipe);
+        eframe::set_value(storage, "RECIPE_SEARCH_TEXT", &self.recipe_search_text);
+        eframe::set_value(storage, "FOOD_SEARCH_TEXT", &self.food_search_text);
+        eframe::set_value(storage, "POTION_SEARCH_TEXT", &self.potion_search_text);
+    }
+
+    fn auto_save_interval(&self) -> std::time::Duration {
+        std::time::Duration::from_secs(1)
+    }
 }
 
 impl MacroSolverApp {
+    fn solver_update(&mut self) {
+        if cfg!(not(target_arch = "wasm32")) {
+            self.solver_update_native_hook();
+        }
+
+        if let Some(update) = self.data_update.take() {
+            match update {
+                SolverEvent::Progress(progress) => {
+                    self.solver_progress = progress;
+                }
+                SolverEvent::IntermediateSolution(actions) => {
+                    self.actions = actions;
+                }
+                SolverEvent::FinalSolution(actions) => {
+                    self.actions = actions;
+                    self.duration = Some(Instant::now() - self.start_time.unwrap());
+                    self.solver_pending = false;
+                }
+            }
+        }
+    }
+
+    #[cfg(target_arch = "wasm32")]
+    fn solver_update_native_hook(&mut self) {
+        // This function does nothing on wasm32 targets.
+    }
+
+    // This could be optimized. But for ease of maintainability, this is most likely easier
+    #[cfg(not(target_arch = "wasm32"))]
+    fn solver_update_native_hook(&mut self) {
```

Comment:
This could be inlined to avoid the empty `solver_update_native_hook` function for `wasm`.

Reply:
Fixed in https://github.com/Human-Hamster/raphael-rs/commit/a4c1bada60ed6429062f10cb452d10a8957f90d7

Reply:
Actually I now remembered that I had originally inlined this. But had to separate because it doesn't compile on WASM.

```
error[E0609]: no field `rx` on type `WorkerBridge<worker::Worker>`
   --> src\app.rs:356:51
    |
356 |             if let Some(bridge_rx) = &self.bridge.rx {
    |                                                   ^^ unknown field
```

Reply:
I can probably get around this altogether, by passing the data update cell pointer to the bridge during construction. So not a problem.

Reply:
You could also solve it by using `#[cfg(not(target_arch = "wasm32"))]` instead of `if cfg!`, like so:
```rust
#[cfg(not(target_arch = "wasm32"))]
if let Some(bridge_rx) = &self.bridge.rx {
    if let Ok(update) = bridge_rx.try_recv() {
        self.data_update.set(Some(update));
    }
}
```
This way, the entire if statement is not visible to the compiler on wasm.

Reply:
I had forgotten you could do this. Thanks!